File: android/src/main/java/expo/modules/superwallexpo/SuperwallExpoModule.kt
package expo.modules.superwallexpo

import android.content.pm.PackageManager
import android.net.Uri
import android.app.Application
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.Promise
import expo.modules.superwallexpo.bridges.PurchaseControllerBridge
import expo.modules.superwallexpo.bridges.SuperwallDelegateBridge
import expo.modules.superwallexpo.json.*
import com.superwall.sdk.Superwall
import com.superwall.sdk.delegate.SuperwallDelegate
import com.superwall.sdk.delegate.PurchaseResult
import com.superwall.sdk.delegate.RestorationResult
import com.superwall.sdk.models.entitlements.SubscriptionStatus
import com.superwall.sdk.models.entitlements.Entitlement
import com.superwall.sdk.identity.IdentityOptions
import com.superwall.sdk.identity.identify
import com.superwall.sdk.identity.setUserAttributes
import com.superwall.sdk.config.options.SuperwallOptions
import com.superwall.sdk.logger.LogLevel
import com.superwall.sdk.network.device.InterfaceStyle
import com.superwall.sdk.paywall.presentation.PaywallPresentationHandler
import com.superwall.sdk.paywall.presentation.result.PresentationResult
import com.superwall.sdk.config.models.ConfigurationStatus
import com.superwall.sdk.paywall.presentation.register
import com.superwall.sdk.paywall.presentation.dismiss
import com.superwall.sdk.paywall.presentation.get_presentation_result.getPresentationResult
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import android.util.Log

class SuperwallExpoModule : Module() {

  companion object {
    var instance: SuperwallExpoModule? = null
  }

  val scope = CoroutineScope(Dispatchers.Main)
  val ioScope = CoroutineScope(Dispatchers.IO)
  init {
    instance = this
  }

  fun emitEvent(name: String, body: Map<String, Any?>?) {
    SuperwallExpoModule.instance?.sendEvent(name, body?:emptyMap())
  }

  private val onPaywallPresent = "onPaywallPresent"
  private val onPaywallDismiss = "onPaywallDismiss"
  private val onPaywallError = "onPaywallError"
  private val onPaywallSkip = "onPaywallSkip"

  // Purchase Events
  private val onPurchase = "onPurchase"
  private val onPurchaseRestore = "onPurchaseRestore"

  // Legacy Events
  private val subscriptionStatusDidChange = "subscriptionStatusDidChange"
  private val handleSuperwallEvent = "handleSuperwallEvent"
  private val handleCustomPaywallAction = "handleCustomPaywallAction"
  private val willDismissPaywall = "willDismissPaywall"
  private val willPresentPaywall = "willPresentPaywall"
  private val didDismissPaywall = "didDismissPaywall"
  private val didPresentPaywall = "didPresentPaywall"
  private val paywallWillOpenURL = "paywallWillOpenURL"
  private val paywallWillOpenDeepLink = "paywallWillOpenDeepLink"
  private val handleLog = "handleLog"
  private val willRedeemLink = "willRedeemLink"
  private val didRedeemLink = "didRedeemLink"
  val purchaseController = PurchaseControllerBridge.instance

  override fun definition() = ModuleDefinition {


    Name("SuperwallExpo")

    Events(
      onPaywallPresent,
      onPaywallDismiss,
      onPaywallError,
      onPaywallSkip,
      onPurchase,
      onPurchaseRestore,

      // Legacy events
      subscriptionStatusDidChange,
      handleSuperwallEvent,
      handleCustomPaywallAction,
      willDismissPaywall,
      willPresentPaywall,
      didDismissPaywall,
      didPresentPaywall,
      paywallWillOpenURL,
      paywallWillOpenDeepLink,
      handleLog,
      willRedeemLink,
      didRedeemLink
    )

    Function("getApiKey") {
      val applicationInfo = appContext?.reactContext?.packageManager?.getApplicationInfo(appContext?.reactContext?.packageName.toString(), PackageManager.GET_META_DATA)
      return@Function applicationInfo?.metaData?.getString("SUPERWALL_API_KEY")
    }

    Function("identify") { userId: String, options: Map<String, Any> ->
      try {
        val options = identityOptionsFromJson(options)
        Superwall.instance.identify(userId = userId, options = options)
      } catch (error: Exception) {
        error.printStackTrace()
      }
    }

    Function("reset") {
      Superwall.instance.reset()
    }

    AsyncFunction("configure") {
      apiKey: String,
      options: Map<String, Any>?,
      usingPurchaseController: Boolean?,
      sdkVersion: String?,
      promise: Promise ->
      //TODO SDK version in arguments?
      try{
      val superwallOptions: SuperwallOptions = options?.let {
        superwallOptionsFromJson(options)
      }?:SuperwallOptions()


      Superwall.configure(
        apiKey = apiKey,
        applicationContext = appContext.reactContext?.applicationContext as Application,
        purchaseController = if (usingPurchaseController == true) purchaseController else null,
        activityProvider = ExpoActivityProvider(appContext),
        options = superwallOptions,
        completion = {
          Superwall.instance.setPlatformWrapper("Expo", version = sdkVersion ?: "0.0.0")
          Superwall.instance.delegate = SuperwallDelegateBridge()
          promise.resolve(true)
         }
       )
      } catch (error: Throwable) {
        error.printStackTrace()
        promise.reject(CodedException(error))
      }
    }

    AsyncFunction("getConfigurationStatus") { promise: Promise ->
      try {
        val configurationStatus = Superwall.instance.configurationState.asString()
          promise.resolve(configurationStatus)
        } catch (error: Throwable) {
          promise.reject(CodedException(error))
        }
      }

    AsyncFunction("registerPlacement") {
      placement: String,
      params: Map<String, Any>?,
      handlerId: String?,
      promise: Promise ->



      var handler: PaywallPresentationHandler? = handlerId?.let {
        PaywallPresentationHandler()
      }



        handler?.onPresent { paywallInfo ->
          val data =
          mapOf(
            "paywallInfoJson" to paywallInfo.toJson(),
          "handlerId" to handlerId)
          sendEvent(onPaywallPresent, data)
        }



        handler?.onDismiss { paywallInfo, result ->
          val data =
          mapOf(
            "paywallInfoJson" to paywallInfo.toJson(),
          "result" to result.toJson(),
          "handlerId" to handlerId)
          
          sendEvent(onPaywallDismiss, data)
        }

        handler?.onError { error ->
          val data =
          mapOf(
            "errorString" to error.localizedMessage,
          "handlerId" to handlerId,
          )
          sendEvent(onPaywallError, data)
        }

        handler?.onSkip { reason ->
          val data =
          mapOf(
            "skippedReason" to reason.toJson(),
             "handlerId" to handlerId)
          sendEvent(onPaywallSkip, data)
        }

      Superwall.instance.register(
        placement = placement,
        params = params,
        handler = handler
      ) {
        promise.resolve(null)
      }
    }
    

    AsyncFunction("getAssignments") { promise: Promise ->
        Superwall.instance.getAssignments()
        .fold({
          promise.resolve(it.map { it.toJson() })
        }, { error ->
          promise.reject(CodedException(error))
        })
      }

    AsyncFunction("getEntitlements") { promise: Promise ->
      try {
        val entitlements = Superwall.instance.entitlements
        val map = mutableMapOf<String, List<Map<String, String>>>()
        
        map["all"] = entitlements.all.map { entitlement ->
          mapOf("id" to entitlement.id)
        }
        
        map["inactive"] = entitlements.inactive.map { entitlement ->
          mapOf("id" to entitlement.id)
        }
        
        map["active"] = entitlements.active.map { entitlement ->
          mapOf("id" to entitlement.id)
        }
        promise.resolve(map)
      } catch (error: Exception) {
        promise.reject(CodedException(error))
      }
      }

    AsyncFunction("getSubscriptionStatus") { promise: Promise ->
      try {
        val subscriptionStatus = Superwall.instance.subscriptionStatus.value.toJson()
          promise.resolve(subscriptionStatus)
      } catch (error: Exception) {
        promise.reject(CodedException(error))
      }
    }

    Function("setSubscriptionStatus") { status: Map<String,Any> ->
      val statusString = (status["status"] as? String)?.uppercase() ?: "UNKNOWN"
      val subscriptionStatus: SubscriptionStatus

      when (statusString) {
        "UNKNOWN" -> subscriptionStatus = SubscriptionStatus.Unknown
        "INACTIVE" -> subscriptionStatus = SubscriptionStatus.Inactive
        "ACTIVE" -> {
          val entitlementsArray = status["entitlements"] as? List<Map<String, Any>>
          val entitlementsSet: Set<Entitlement> = entitlementsArray?.map { item ->
            val id = item["id"] as? String
            id?.let { Entitlement(id = it) }
          }?.filterNotNull()?.toSet() ?: emptySet()
          subscriptionStatus = SubscriptionStatus.Active(entitlementsSet)
        }
        else -> subscriptionStatus = SubscriptionStatus.Unknown
      }

      Superwall.instance.setSubscriptionStatus(subscriptionStatus)
    }
    
    Function("setInterfaceStyle") { style: String? ->
      var interfaceStyle: InterfaceStyle? = style?.let { interfaceStyleFromString(it) }
      Superwall.instance.setInterfaceStyle(interfaceStyle)
    }

    AsyncFunction("getUserAttributes") { promise: Promise ->
      val attributes = Superwall.instance.userAttributes
      promise.resolve(attributes)
    }

    Function("setUserAttributes") { userAttributes: Map<String, Any> ->
      Superwall.instance.setUserAttributes(userAttributes)
    }

    AsyncFunction("handleDeepLink") { url: String, promise: Promise ->
      val url = Uri.parse(url)
      val result = Superwall.instance.handleDeepLink(url).fold({
        promise.resolve(it)
      }, { error ->
        promise.reject(CodedException(error))
      })
    }

    Function("didPurchase") { result: Map<String, Any> ->
      val purchaseResult = purchaseResultFromJson(result)
      purchaseController.purchasePromise?.complete(purchaseResult)
    }

    Function("didRestore") { result: Map<String, Any> ->
      val restorationResult = restorationResultFromJson(result)
      purchaseController.restorePromise?.complete(restorationResult)
    }

    AsyncFunction("dismiss") { promise: Promise ->
      ioScope.launch {
        Superwall.instance.dismiss()
        scope.launch {
          promise.resolve(null)
        }
      }
    }

    AsyncFunction("confirmAllAssignments") { promise: Promise ->
      ioScope.launch {
        Superwall.instance.confirmAllAssignments().fold({
          scope.launch {
            promise.resolve(it.map { it.toJson() })
          }
        }, { error ->
          scope.launch {
            promise.reject(CodedException(error))
          }
        })
      }
    }

    AsyncFunction("getPresentationResult") { placement: String, params: Map<String, Any>?, promise: Promise ->
      ioScope.launch {
        Superwall.instance.getPresentationResult(placement = placement, params = params)
        .fold({ result ->
          scope.launch {
            promise.resolve(result.toJson())
          }
        }, { error ->
          scope.launch {
            promise.reject(CodedException(error))
          }
        })
      }
    }

    Function("preloadPaywalls") { placementNames: List<String> ->
      Superwall.instance.preloadPaywalls(placementNames = placementNames.toSet())
    }

    Function("preloadAllPaywalls") {
      Superwall.instance.preloadAllPaywalls()
    }

    Function("setLogLevel") { level: String ->
      val logLevel = LogLevel.values().find { it.toString().equals(level, ignoreCase = true) }
      if (logLevel != null) {
        Superwall.instance.logLevel = logLevel
      }
    }
  }
}

File: ios/SuperwallExpoModule.swift
import ExpoModulesCore
import SuperwallKit

public class SuperwallExpoModule: Module {
  public static var shared: SuperwallExpoModule?

  private let purchaseController = PurchaseControllerBridge.shared

  // Paywall Events
  private let onPaywallPresent = "onPaywallPresent"
  private let onPaywallDismiss = "onPaywallDismiss"
  private let onPaywallError = "onPaywallError"
  private let onPaywallSkip = "onPaywallSkip"

  // Purchase Events
  private let onPurchase = "onPurchase"
  private let onPurchaseRestore = "onPurchaseRestore"

  // Corrected initializer
  public required init(appContext: AppContext) {
    super.init(appContext: appContext)
    SuperwallExpoModule.shared = self
  }

  public static func emitEvent(_ name: String, _ data: [String: Any]?) {
    SuperwallExpoModule.shared?.sendEvent(name, data ?? [:])
  }

  public func definition() -> ModuleDefinition {
    Name("SuperwallExpo")

    Events(
      onPaywallPresent,
      onPaywallDismiss,
      onPaywallError,
      onPaywallSkip,
      onPurchase,
      onPurchaseRestore
    )

    Function("getApiKey") {
      return Bundle.main.object(forInfoDictionaryKey: "SUPERWALL_API_KEY")
        as? String
    }

    Function("identify") {
      (userId: String, options: [String: Any]?) in
      let options = IdentityOptions.fromJson(options)
      Superwall.shared.identify(userId: userId, options: options)
    }

    Function("reset") {
      Superwall.shared.reset()
    }

    // Function("setDelegate") {
    //   (isUndefined: Bool) in
    //   self.delegate = isUndefined ? nil : SuperwallDelegateBridge()
    //   Superwall.shared.delegate = self.delegate
    // }

    AsyncFunction("configure") {
      (
        apiKey: String,
        options: [String: Any]?,
        usingPurchaseController: Bool,
        sdkVersion: String,
        promise: Promise
      ) in

      var superwallOptions: SuperwallOptions?

      if let options = options {
        superwallOptions = SuperwallOptions.fromJson(options)
      }

      print("SuperwallExpoModule.configure()")
      print("apiKey: \(apiKey)")

      Superwall.configure(
        apiKey: apiKey,
        purchaseController: usingPurchaseController ? purchaseController : nil,
        options: superwallOptions
      ) {
        promise.resolve(nil)
      }

      Superwall.shared.setPlatformWrapper("React Native", version: sdkVersion)
    }

    AsyncFunction("getConfigurationStatus") { (promise: Promise) in
      do {
        let configurationStatus = try Superwall.shared.configurationStatus.toString()
        promise.resolve(configurationStatus)
      } catch {
        promise.reject(error)
      }
    }

    AsyncFunction("registerPlacement") {
      (
        placement: String,
        params: [String: Any]?,
        handlerId: String?,
        promise: Promise
      ) in
      var handler: PaywallPresentationHandler?

      if let handlerId = handlerId {
        handler = PaywallPresentationHandler()

        handler?.onPresent { [weak self] paywallInfo in
          guard let self = self else { return }
          let data =
            [
              "paywallInfoJson": paywallInfo.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallPresent, data)
        }

        handler?.onDismiss { [weak self] paywallInfo, result in
          guard let self = self else { return }
          let data =
            [
              "paywallInfoJson": paywallInfo.toJson(),
              "result": result.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallDismiss, data)
        }

        handler?.onError { [weak self] error in
          guard let self = self else { return }
          let data =
            [
              "errorString": error.localizedDescription,
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallError, data)
        }

        handler?.onSkip { [weak self] reason in
          guard let self = self else { return }
          let data =
            [
              "skippedReason": reason.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallSkip, data)
        }
      }

      Superwall.shared.register(
        placement: placement,
        params: params,
        handler: handler
      ) {
        promise.resolve(nil)
      }
    }

    AsyncFunction("getAssignments") {
      (promise: Promise) in
      do {
        let assignments = try Superwall.shared.getAssignments()
        promise.resolve(assignments.map { $0.toJson() })
      } catch {
        promise.reject(error)
      }
    }

    AsyncFunction("getEntitlements") {
      (promise: Promise) in
      do {
        let entitlements = try Superwall.shared.entitlements.toJson()
        promise.resolve(entitlements)
      } catch {
        promise.reject(error)
      }
    }

    AsyncFunction("getSubscriptionStatus") {
      (promise: Promise) in
      let subscriptionStatus = Superwall.shared.subscriptionStatus
      promise.resolve(subscriptionStatus)
    }

    Function("setSubscriptionStatus") {
      (status: [String: Any]) in
      let statusString = (status["status"] as? String)?.uppercased() ?? "UNKNOWN"
      let subscriptionStatus: SubscriptionStatus

      switch statusString {
      case "UNKNOWN":
        subscriptionStatus = .unknown
      case "INACTIVE":
        subscriptionStatus = .inactive
      case "ACTIVE":
        if let entitlementsArray = status["entitlements"] as? [[String: Any]] {
          let entitlementsSet: Set<Entitlement> = Set(
            entitlementsArray.compactMap { item in
              if let id = item["id"] as? String {
                return Entitlement(id: id)
              }
              return nil
            }
          )
          subscriptionStatus = .active(entitlementsSet)
        } else {
          subscriptionStatus = .inactive
        }
      default:
        subscriptionStatus = .unknown
      }

      Superwall.shared.subscriptionStatus = subscriptionStatus
    }
  }
}
